Add developer instructions for new templates (#564)
XanaduAI/pennylane: doc/development/adding_templates.rst

   subdirectory. The file contains your new template, which is a function that defines a sequence of
   quantum gates (without measurements).

   For example, this is a very basic template applying an :class:`~.RX`
   gate to each wire:

   .. code-block:: python

        import pennylane as qml
        from pennylane.templates import template  # import the decorator

        @template
        def MyNewTemplate(weights, wires):

            # Check that the inputs have the correct format
            # ...

            for wire, weight in zip(wires, weights):
                qml.RX(weight, wires=wire)

   Since the template is called within a :ref:`quantum function <intro_vcirc_qfunc>`,
   it can only contain information processing that is allowed
   inside a quantum functions.

   Make sure you consider the following:

   * *Choose the name carefully.* Good names tell the user what a template is good for,
      or what architecture it implements. To be consistent with other quantum operations
      (which are classes), the function name (i.e., ``MyNewTemplate``) is written in camel case. 

>    * *Register the template.* A template is "registered" by using the :mod:`template <pennylane.template>`

Reviewer: <pennylane.template**s**> ??

Author: I was wondering for a second too, but this is not referring to the ``templates`` module, but to the function ``template``. Good observation though.

     decorator ``@template``. This allows us to record the queue of operations of a template,
     which is very useful for testing:

     .. code-block:: python

        with qml.utils.OperationRecorder() as rec:
            MyNewTemplate(...)

        list_of_gates = rec.queue

>    * Consider using the :func:`broadcasting <pennylane.templates.broadcast.broadcast>` function to make your

Reviewer 2: My bad with the previous suggestion, need to change the name as well. Also, the link seems to work in a slightly modified way for some reason :thinking: 
```suggestion
   * Consider using the :func:`broadcast <pennylane.broadcast>` function to make your
```

Reviewer 3: How about just
```suggestion
   * Consider using the :func:`~.broadcast` function to make your
```
😆

>      code more concise.

Reviewer: nice!

     by using different interfaces (i.e., an input could be a ``numpy.ndarray`` or a list of
     :class:`Variable <pennylane.variable.Variable>`, depending on how the user uses the template).

2. **Import the new template** in ``templates/<templ_type>/__init__.py`` by adding the new line

   .. code-block:: python

        from .mynewtemplate import MyNewTemplate

3. **Add your template to the documentation** by adding a ``customgalleryitem`` to the correct layer type section in
   ``doc/introduction/templates.rst``:

   .. code-block::

     .. customgalleryitem::
        :link: ../code/api/pennylane.templates.<templ_type>.MyNewTemplate.html
        :description: MyNewTemplate
        :figure: ../_static/templates/<templ_type>/my_new_template.png

   .. note::

      This loads the image of the template added to ``doc/_static/templates/<templ_type>/`` in Step 1. Make sure that
      this image has the same dimensions and style as other template icons in the folder.

4. **Add tests** for your new template to the test suite.

   * Integration tests, which check that your template can be called inside a quantum node and that PennyLane can

>      compute gradients with respect to differentiable parameters, are added to ``tests/test_templates.py``.

Reviewer 2: (Has the downside of being an absolute link, susceptible to any changes)
```suggestion
     compute gradients with respect to differentiable parameters, are added to `tests/test_templates.py <https://github.com/XanaduAI/pennylane/blob/master/tests/test_templates.py>`_.
```

How to add a new template
-------------------------

The following steps will help you to add your favourite circuit ansatz to
PennyLane's :mod:`template <pennylane.template>` library.

PennyLane internally distinguishes :ref:`different types <intro_ref_temp>` of templates, such as
:mod:`embeddings <pennylane.templates.embeddings>`, :mod:`layers <pennylane.templates.layers>`,
:mod:`state_preparations <pennylane.templates.state_preparations>` or
:mod:`soubroutines <pennylane.templates.soubroutines>`.
Here we will use the template ``MyNewTemplate`` as an example, and you need to replace ``<templ_type>`` with the
correct template type.

1. **Add the template** by adding a new file ``my_new_template.py`` to the correct ``templates/<templ_type>/``

>    subdirectory, which contains your new template of the form

Reviewer 3: Perhaps it might make it might be easier to follow if the guide walked through the process of adding an actual template example. For example,

```python
@template
def MyLayer(weights, wires, entangler=qml.CZ):
    # The template can apply arbitrary quantum operations.
    # The operations applied may depend on the shape of input arguments.

   # iterate through layers
    for layer, p in enumerate(weights):
        for w in wires:
            # apply X rotation to each wire
            qml.RX(w, wires=w)

        # The qml.broadcast function may also be used.
        # Apply entangler to wire pairs that connect all wires to each other.
        qml.broadcast(entangler, wires, pattern="all_to_all")
```

Reviewer 3: The steps below could be made more clear by going back to this example, e.g., "In this example, since we are creating a layer template, we place this function in the file `pennylane/templates/layers/my_layer.py`"

Author: That is a nice idea. However, I substantially simplified the example, as this one has a lot of subtleties...

Reviewer 3: Feel free!

   .. code-block:: python

        @template
        def MyNewTemplate(*args, **kwargs):
            ...

   A template is a function that defines a sequence of quantum gates (without measurements).
   Since the template is called within a :ref:`quantum function <intro_vcirc_qfunc>`,
   it can only contain information processing that is allowed
   inside a quantum functions.

   Make sure you consider the following:

   * To be consistent with other quantum operations (which are classes), the function name (i.e., ``MyNewTemplate``) is

>      capitalized. Choose the name carefully. Good names tell the user what a template is good for, or what architecture

Reviewer 3: ```suggestion
     written in camel case. Choose the name carefully. Good names tell the user what a template is good for, or what architecture
```

     it implements.

   * A template is "registered" by using the :mod:`template <pennylane.template>` decorator ``@template``.
     This allows us to record the queue of operations of a template, which is very useful for testing:

     .. code-block:: python

        with qml.utils.OperationRecorder() as rec:
            MyNewTemplate(...)

        list_of_gates = rec.queue

   * Your template should have a positional argument called ``wires``, which takes an integer (single wire) or a list

>      of wire indices that the template acts on.

Reviewer 3: An example here would be good

Author: Actually, I am so unsure if this is even true. In principle, one does not need this argument, but in the past it has been a convention. Deleted it for now.

   * Consider using the :func:`broadcasting <pennylane.templates.broadcasting>` function to make your code more concise.

   * Write an extensive docstring that explains how to use the template. Include a sketch of the template (add the
     file to the ``doc/_static/templates/<templ_type>/`` directory). You can also display a small usage example
     at the beginning of the docstring.

     At the end of the docstring, add a section starting with the ``.. UsageDetails::`` directive,
     where you demonstrate with code examples how to use the templates with different

>      settings, for example varying the number of wires, explaining keyword arguments and special cases.

Reviewer 3: Could be good to include a link to existing docstrings for reference!

   * Check the inputs to the template. You can use the functions provided in :mod:`utils <pennylane.templates.utils>`.
     Don't forget that arguments may be passed by the user to the qnode as positional or keyword arguments, and
     by using different interfaces (i.e., a input could be a ``numpy.ndarray`` or a list of

>      :class:`Variable <pennylane.variable.Variable>`, depending on how the user uses the template).

Reviewer 3: I've just noticed that the pennylane.templates.utils module is not linked to on the templates code page (https://pennylane.readthedocs.io/en/stable/code/qml_templates.html). It might be good to include it (somewhere) so that the documentation is available (as it is, I would have no idea what is available/how they're used without looking in the source code directly)

Author: Good point. Do you know how to use ``autosummary`` so that all functions from a file are listed (without that file being a module?)

Reviewer 3: Yep! See https://pennylane.readthedocs.io/en/stable/development/guide.html#adding-a-new-module-to-the-docs

Author: Lol that was in the file I was adding to. I should have read it!

Author: I see, ``automodapi`` is what I tried before, and it does not import the functions. From playing around I assume that this is because they are private (i.e. they start with underscore). Since developers use them they are kind of semi-private.

Shall I refactor them to be public? Or resolve both a separate PR?

Author: Alternatively, I could list all functions manually by using ``autosummary``

Author: Ok, I solved it for now by using ``automodapi``, but the list in the docs (https://pennylane.readthedocs.io/en/stable/code/qml_templates.html) will be empty for now. I am preparing a separate PR that refactors the utils functions to be non-private. This is consistent with other utils modules in PennyLane...

2. **Import the new template** in ``templates/<templ_type>/__init__.py`` by adding the new line

   .. code-block:: python

        from .mynewtemplate import MyNewTemplate

3. **Add your template to the documentation** by adding a ``customgalleryitem`` to the correct layer type section in
   ``doc/introduction/templates.rst``:

   .. code-block::

     .. customgalleryitem::
        :link: ../code/api/pennylane.templates.<templ_type>.MyNewTemplate.html
        :description: MyNewTemplate
        :figure: ../_static/templates/<templ_type>/my_new_template.png

   .. note::

      This loads the image of the template added to ``doc/_static/templates`` in Step 1. Make sure that
      this image has the same dimensions and style than other template icons.

4. **Add tests** for your new template to the test suite.

   * Integration tests which check that your template can be called inside a quantum node, and that PennyLane can
     compute gradients with respect to differentiable parameters, are added to ``tests/test_templates.py``.
     Simply add your template to the fixtures (variables indicated by capital letters) to automatically

>      run existing tests on your new template.

Reviewer 3: This may need more explanation 😆

Author: I was thinking to add the explanation to the test file, so it is easier to maintain?

Author: [I thought I replied, but seems it didn't go through...]

For easy maintenance I'd love to put the explanation into the test file itself!

How to add a new template
-------------------------

The following steps will help you to add your favourite circuit ansatz to
PennyLane's :mod:`template <pennylane.template>` library.

PennyLane internally distinguishes :ref:`different types <intro_ref_temp>` of templates, such as
:mod:`embeddings <pennylane.templates.embeddings>`, :mod:`layers <pennylane.templates.layers>`,
:mod:`state_preparations <pennylane.templates.state_preparations>` or

> :mod:`soubroutines <pennylane.templates.soubroutines>`.

Reviewer 2: ```suggestion
:mod:`subroutines <pennylane.templates.subroutines>`.
```

        @template
        def MyNewTemplate(weights, wires):

            # Check that the inputs have the correct format
            # ...

            for wire, weight in zip(wires, weights):
                qml.RX(weight, wires=wire)

   A template is a function that defines a sequence of quantum gates (without measurements).
   Since the template is called within a :ref:`quantum function <intro_vcirc_qfunc>`,
   it can only contain information processing that is allowed
   inside a quantum functions.

   Make sure you consider the following:

   * To be consistent with other quantum operations (which are classes), the function name (i.e., ``MyNewTemplate``) is
     written in camel case. Choose the name carefully. Good names tell the user what a template is good for, or what architecture
     it implements.

   * A template is "registered" by using the :mod:`template <pennylane.template>` decorator ``@template``.
     This allows us to record the queue of operations of a template, which is very useful for testing:

     .. code-block:: python

        with qml.utils.OperationRecorder() as rec:
            MyNewTemplate(...)

        list_of_gates = rec.queue

>    * Consider using the :func:`broadcasting <pennylane.templates.broadcasting>` function to make your code more concise.

Reviewer 2: ```suggestion
   * Consider using the :func:`broadcasting <pennylane.templates.broadcast>` function to make your code more concise.
```

   * Write an extensive docstring that explains how to use the template. Include a sketch of the template (add the
     file to the ``doc/_static/templates/<templ_type>/`` directory). You can also display a small usage example
     at the beginning of the docstring.

     At the end of the docstring, add a section starting with the ``.. UsageDetails::`` directive,
     where you demonstrate with code examples how to use the templates with different
     settings, for example varying the number of wires, explaining keyword arguments and special cases.
     For inspiration, check one of the existing templates, such as

>      :func:`AmplitudeEmbedding <pennylane.templates.embeddings.AmplitudeEmbedding>`

Reviewer 2: ```suggestion
     :func:`AmplitudeEmbedding <pennylane.templates.embeddings.AmplitudeEmbedding>`.
```

   * Check the inputs to the template. You can use the functions provided in :mod:`utils <pennylane.templates.utils>`.

>      Don't forget that arguments may be passed by the user to the qnode as positional or keyword arguments, and

Reviewer 2: I'm assuming that this was originally meant to refer to differentiability of arguments in PL. Something like the suggestion (though that might be the very first time a user sees these terms) or explicitly stating the difference beforehand could work well I guess.
```suggestion
     Don't forget that arguments may be passed by the user to the qnode as primary or auxiliary quantum function arguments, and
```

>      by using different interfaces (i.e., a input could be a ``numpy.ndarray`` or a list of

Reviewer 2: ```suggestion
     by using different interfaces (i.e., an input could be a ``numpy.ndarray`` or a list of
```

     :class:`Variable <pennylane.variable.Variable>`, depending on how the user uses the template).

2. **Import the new template** in ``templates/<templ_type>/__init__.py`` by adding the new line

   .. code-block:: python

        from .mynewtemplate import MyNewTemplate

3. **Add your template to the documentation** by adding a ``customgalleryitem`` to the correct layer type section in
   ``doc/introduction/templates.rst``:

   .. code-block::

     .. customgalleryitem::
        :link: ../code/api/pennylane.templates.<templ_type>.MyNewTemplate.html
        :description: MyNewTemplate
        :figure: ../_static/templates/<templ_type>/my_new_template.png

   .. note::

      This loads the image of the template added to ``doc/_static/templates`` in Step 1. Make sure that

>       this image has the same dimensions and style than other template icons.

Reviewer 2: ```suggestion
      this image has the same dimensions and style as other template icons.
```
